File: src/main/java/org/example/bamsemats/dotatranslator/TranslationService.java
package org.example.bamsemats.dotatranslator;

import com.google.api.client.auth.oauth2.Credential;
import com.google.api.gax.core.FixedCredentialsProvider;
import com.google.auth.oauth2.UserCredentials; // Change import from GoogleCredentials to UserCredentials
import com.google.auth.oauth2.AccessToken; // Correctly placed import

import java.io.IOException;

import com.google.cloud.translate.v3.*;
import javafx.application.Platform;
import java.util.HashMap;
import java.util.Map;
import java.util.Date; // Needed for expiration time

public class TranslationService {

    private final String projectId;
    private final TranslationServiceClient client;
    private final MainViewController mainViewController;
    private final Map<String, String> translationCache = new HashMap<>();

    private static final String TARGET_LANGUAGE = "en";

    // Constructor now accepts a Credential, Client ID, and Client Secret
    public TranslationService(String projectId, Credential credential, String clientId, String clientSecret, MainViewController mainViewController) throws IOException {
        this.projectId = projectId;
        this.mainViewController = mainViewController;
        TranslationServiceClient tempClient = null;
        if (credential != null && credential.getAccessToken() != null) { // Check if credential and its token are valid
            try {
                // Create com.google.auth.oauth2.AccessToken from com.google.api.client.auth.oauth2.Credential
                com.google.auth.oauth2.AccessToken googleAuthAccessToken = new com.google.auth.oauth2.AccessToken(
                    credential.getAccessToken(),
                    new Date(credential.getExpirationTimeMilliseconds())
                );

                // Create UserCredentials from the com.google.api.client.auth.oauth2.Credential
                UserCredentials.Builder userCredentialsBuilder = UserCredentials.newBuilder()
                    .setAccessToken(googleAuthAccessToken)
                    .setClientId(clientId) // Set client ID
                    .setClientSecret(clientSecret); // Set client secret

                if (credential.getRefreshToken() != null) {
                    userCredentialsBuilder.setRefreshToken(credential.getRefreshToken());
                }
                
                UserCredentials credentials = userCredentialsBuilder.build();
                
                TranslationServiceSettings settings = TranslationServiceSettings.newBuilder()
                    .setCredentialsProvider(FixedCredentialsProvider.create(credentials))
                    .build();
                tempClient = TranslationServiceClient.create(settings);
            } catch (Exception e) {
                Platform.runLater(() -> mainViewController.addMessage("Error initializing Google Cloud Translation Service with Credential: " + e.getClass().getSimpleName() + ". Translation will be disabled."));
                System.err.println("Error initializing Google Cloud Translation Service with Credential:");
                e.printStackTrace();
            }
        } else {
            Platform.runLater(() -> mainViewController.addMessage("No Google Cloud Credential (or valid Access Token) provided. Translation will be disabled."));
        }
        this.client = tempClient;
    }

    public String translateText(String sourceText, String sourceLanguage) {
        if (this.client == null || sourceText == null || sourceText.trim().isEmpty() || TARGET_LANGUAGE.equalsIgnoreCase(sourceLanguage)) {
            return sourceText;
        }

        String cacheKey = sourceText + "::" + sourceLanguage;
        if (translationCache.containsKey(cacheKey)) {
            return translationCache.get(cacheKey);
        }

        if (UsageTracker.isTranslationLimitReached()) {
            Platform.runLater(() -> mainViewController.addMessage("Warning: Translation free tier limit reached for this month. Further translation requests are blocked."));
            return sourceText;
        }

        try {
            LocationName parent = LocationName.of(projectId, "global");

            TranslateTextRequest request =
                    TranslateTextRequest.newBuilder()
                            .setParent(parent.toString())
                            .setMimeType("text/plain")
                            .setTargetLanguageCode(TARGET_LANGUAGE)
                            .addContents(sourceText)
                            .setSourceLanguageCode(sourceLanguage)
                            .build();

            TranslateTextResponse response = client.translateText(request);

            UsageTracker.incrementTranslationCharacters(sourceText.length());

            if (UsageTracker.getTranslationUsagePercentage() >= 80 && UsageTracker.getTranslationUsagePercentage() < 100) {
                Platform.runLater(() -> mainViewController.addMessage(
                        String.format("Warning: Translation usage is at %.0f%% of the free tier limit (%d/%d characters). You may be billed soon.",
                                UsageTracker.getTranslationUsagePercentage(), UsageTracker.getTranslationCharacters(), UsageTracker.getTranslationFreeTierLimit())
                ));
            }

            if (!response.getTranslationsList().isEmpty()) {
                String translatedText = response.getTranslationsList().get(0).getTranslatedText();
                translationCache.put(cacheKey, translatedText);
                return translatedText;
            }
        } catch (Exception e) {
            System.err.println("Error during translation: " + e.getMessage());
            Platform.runLater(() -> mainViewController.addMessage("Error translating: " + e.getMessage()));
        }
        return sourceText;
    }

    public void shutdown() {
        if (client != null) {
            client.close();
        }
    }
}

File: translation_service.py
from google.cloud import translate_v3 as translate
from usage_tracker import UsageTracker # Import UsageTracker

TARGET_LANGUAGE = "en" # Always translate to English

class TranslationService:
    def __init__(self, project_id):
        self.project_id = project_id
        self.client = None # Will be initialized after OAuth
        self.usage_tracker = UsageTracker() # Instantiate UsageTracker

    def initialize_client(self, credentials):
        """Initializes the Google Cloud Translation client with provided credentials."""
        self.client = translate.TranslationServiceClient(credentials=credentials)

    def translate_text(self, text, source_language="und"):
        """
        Translates text using Google Cloud Translation API.
        :param text: The text to translate.
        :param source_language: The detected source language code (e.g., 'es', 'fr', 'und' for undetermined).
        :return: Translated text.
        """
        original_text = text # Store original text

        if self.client is None:
            print("Translation client not initialized. Cannot perform translation.")
            return original_text, original_text

        if self.usage_tracker.is_translation_limit_reached():
            print("Warning: Translation free tier limit reached for this month. Further translation requests are blocked.")
            return original_text, original_text
        
        if not text.strip() or source_language.lower() == TARGET_LANGUAGE:
            return original_text, original_text # No need to translate empty text or if already target language

        try:
            parent = f"projects/{self.project_id}/locations/global"
            
            # First, attempt language detection if source_language is undetermined
            if source_language == "und":
                lang_detect_response = self.client.detect_language(
                    parent=parent,
                    content=text,
                    mime_type="text/plain"
                )
                if lang_detect_response.languages:
                    source_language = max(lang_detect_response.languages, key=lambda x: x.confidence).language_code
                else:
                    source_language = "en" # Default to English if detection fails

            # Skip translation if source is already target language
            if source_language.lower() == TARGET_LANGUAGE:
                return original_text, original_text

            # Perform translation
            response = self.client.translate_text(
                request={
                    "parent": parent,
                    "contents": [text],
                    "mime_type": "text/plain",
                    "source_language_code": source_language,
                    "target_language_code": TARGET_LANGUAGE,
                }
            )

            self.usage_tracker.increment_translation_characters(len(text))
            if self.usage_tracker.get_translation_usage_percentage() >= 80:
                print(f"Warning: Translation usage is at {self.usage_tracker.get_translation_usage_percentage():.0f}% of the free tier limit ({self.usage_tracker.get_translation_characters()}/{self.usage_tracker.get_translation_free_tier_limit()} characters).")


            if response.translations:
                translated_text = response.translations[0].translated_text
                return original_text, translated_text
            
        except Exception as e:
            print(f"Error during translation: {e}")
        
        return original_text, original_text # Return original text on error
